File: im-desktop/src/services/shared-media-service.ts
import { sharedMediaApi } from '../api/api-client';
import type { SharedMedia, MediaPage, MediaStatistics, LinkPreview, MediaRequest } from '../types/shared-media';

class SharedMediaService {
  async getSharedMedia(request: MediaRequest): Promise<MediaPage> {
    const params: Record<string, string | number> = {
      page: request.page ?? 0,
      size: request.size ?? 20,
    };
    if (request.mediaType) params.mediaType = request.mediaType;
    if (request.senderId) params.senderId = request.senderId;
    if (request.startTime) params.startTime = request.startTime;
    if (request.endTime) params.endTime = request.endTime;
    if (request.keyword) params.keyword = request.keyword;

    return sharedMediaApi.get<MediaPage>(
      `/media/conversation/${request.conversationId}`,
      { params }
    );
  }

  async getMediaTimeline(conversationId: string, page = 0, size = 20): Promise<SharedMedia[]> {
    return sharedMediaApi.get<SharedMedia[]>(
      `/media/timeline/${conversationId}`,
      { params: { page, size } }
    );
  }

  async getMediaStatistics(conversationId: string): Promise<MediaStatistics> {
    return sharedMediaApi.get<MediaStatistics>(
      `/media/statistics/${conversationId}`
    );
  }

  async getSharedLinks(conversationId: string, page = 0, size = 20): Promise<LinkPreview[]> {
    return sharedMediaApi.get<LinkPreview[]>(
      `/media/links/${conversationId}`,
      { params: { page, size } }
    );
  }

  async getAlbumMedia(conversationId: string, albumType: string, page = 0, size = 50): Promise<MediaPage> {
    return sharedMediaApi.get<MediaPage>(
      `/media/album/${conversationId}`,
      { params: { albumType, page, size } }
    );
  }

  async deleteMedia(mediaId: number): Promise<void> {
    return sharedMediaApi.delete(`/media/${mediaId}`);
  }

  formatFileSize(bytes: number): string {
    if (bytes < 1024) return `${bytes} B`;
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
    if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
    return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`;
  }

  getMediaTypeIcon(mediaType: string): string {
    const icons: Record<string, string> = {
      IMAGE: '🖼️', VIDEO: '🎬', AUDIO: '🎵', FILE: '📎', LINK: '🔗', VOICE: '🎤'
    };
    return icons[mediaType] ?? '📄';
  }
}

export const sharedMediaService = new SharedMediaService();

File: im-mobile/lib/services/shared_media_service.dart
import 'package:dio/dio.dart';
import '../models/shared_media.dart';

class SharedMediaService {
  final Dio _dio;

  SharedMediaService(this._dio);

  Future<MediaPage> getSharedMedia({
    required String conversationId,
    String? mediaType,
    int page = 0,
    int size = 20,
    String? senderId,
    int? startTime,
    int? endTime,
  }) async {
    final queryParams = <String, dynamic>{
      'page': page,
      'size': size,
    };
    if (mediaType != null) queryParams['mediaType'] = mediaType;
    if (senderId != null) queryParams['senderId'] = senderId;
    if (startTime != null) queryParams['startTime'] = startTime;
    if (endTime != null) queryParams['endTime'] = endTime;

    final response = await _dio.get(
      '/media/conversation/$conversationId',
      queryParameters: queryParams,
    );
    return MediaPage.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<SharedMedia>> getMediaTimeline(
      String conversationId, {int page = 0, int size = 20}) async {
    final response = await _dio.get(
      '/media/timeline/$conversationId',
      queryParameters: {'page': page, 'size': size},
    );
    return (response.data as List<dynamic>)
        .map((e) => SharedMedia.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<MediaStatistics> getMediaStatistics(String conversationId) async {
    final response = await _dio.get('/media/statistics/$conversationId');
    return MediaStatistics.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<LinkPreview>> getSharedLinks(String conversationId,
      {int page = 0, int size = 20}) async {
    final response = await _dio.get(
      '/media/links/$conversationId',
      queryParameters: {'page': page, 'size': size},
    );
    return (response.data as List<dynamic>)
        .map((e) => LinkPreview.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<MediaPage> getAlbumMedia(
      String conversationId, String albumType,
      {int page = 0, int size = 50}) async {
    final response = await _dio.get(
      '/media/album/$conversationId',
      queryParameters: {'albumType': albumType, 'page': page, 'size': size},
    );
    return MediaPage.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteMedia(int mediaId) async {
    await _dio.delete('/media/$mediaId');
  }

  String formatFileSize(int bytes) {
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1024 * 1024) return '${(bytes / 1024).toStringAsFixed(1)} KB';
    if (bytes < 1024 * 1024 * 1024) {
      return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
    }
    return '${(bytes / (1024 * 1024 * 1024)).toStringAsFixed(2)} GB';
  }

  String getMediaTypeIcon(MediaType type) {
    switch (type) {
      case MediaType.IMAGE: return '🖼️';
      case MediaType.VIDEO: return '🎬';
      case MediaType.AUDIO: return '🎵';
      case MediaType.FILE: return '📎';
      case MediaType.LINK: return '🔗';
      case MediaType.VOICE: return '🎤';
    }
  }
}
